src/sql/src/query_model/hir/hir_from_qgm.rs

```diff
@@ -24,23 +31,393 @@ impl From<Model> for HirRelationExpr {
+    /// `self.common_subgraphs`. Otherwise, it will be pushed to `self.converted`.
+    fn convert_subgraph(&mut self, model: &Model, box_id: &BoxId) {
+        let r#box = model.get_box(*box_id);
+        if self.common_subgraphs.get(&box_id).is_some() {
+            // Subgraph has already been converted.
+            return;
+        }
+        let hir = match &r#box.box_type {
+            BoxType::Get(Get { id, unique_keys }) => {
+                let typ = mz_repr::RelationType::new(
+                    r#box
+                        .columns
+                        .iter()
+                        .map(|c| {
+                            if let BoxScalarExpr::BaseColumn(BaseColumn { column_type, .. }) =
+                                &c.expr
+                            {
+                                column_type.clone()
+                            } else {
+                                // TODO: the validator should make this branch unreachable.
+                                panic!("expected all columns in Get BoxType to be BaseColumn");
+                            }
+                        })
+                        .collect::<Vec<_>>(),
+                )
+                .with_keys(unique_keys.clone());
+                HirRelationExpr::Get {
+                    id: mz_expr::Id::Global(*id),
+                    typ,
+                }
+            }
+            BoxType::Values(values) => HirRelationExpr::constant(
+                // Map `Vec<Vec<BoxScalarExpr>` to `Vec<Vec<Datum>>`
+                values
+                    .rows
+                    .iter()
+                    .map(|row| {
+                        row.iter()
+                            .map(|expr| {
+                                if let BoxScalarExpr::Literal(datum, _) = expr {
+                                    datum.unpack_first()
+                                } else {
+                                    // TODO: the validator should make this branch unreachable.
+                                    panic!("expected all scalars in Values BoxType to be literals");
+                                }
+                            })
+                            .collect_vec()
+                    })
+                    .collect_vec(),
+                mz_repr::RelationType::new(r#box.attributes.get::<BoxRelationType>().to_owned()),
+            ),
+            BoxType::Select(select) => {
+                let (rels, scalars) = self.convert_quantifiers(&r#box);
+                let (join_q_ids, mut join_inputs): (Vec<_>, Vec<_>) = rels.into_iter().unzip();
+                let (map_q_ids, maps): (Vec<_>, Vec<_>) = scalars.into_iter().unzip();
+
+                let mut result = if let Some(first) = join_inputs.pop() {
+                    first
+                } else {
+                    HirRelationExpr::constant(vec![vec![]], mz_repr::RelationType::new(vec![]))
```

Comment:
Not sure if this can ever occur, as a `Select` box without at least one `Foeach` quantifier does not make sense. IMHO the `ConstraintValidator` ensures that a `Select` box must have at least one `Foreach` quantifier and we should return an `Err` here.